fl4health/server/nnunet_server.py

```diff
@@ -56,30 +138,115 @@ def initialize(self, server_round: int, timeout: Optional[float] = None) -> None
-
-        properties = properties_res.properties
-
-        # NnUNetClient has serialized nnunet_plans as a property
-        plans_bytes = properties["nnunet_plans"]
+            # If a checkpoint exists, we load in previously checkpointed values for required properties
+            self.load_server_state()
 
         # Wrap config functions so that nnunet_plans is included
-        new_fit_cfg_fn = add_items_to_config_fn(self.strategy.configure_fit, {"nnunet_plans": plans_bytes})
-        new_eval_cfg_fn = add_items_to_config_fn(self.strategy.configure_evaluate, {"nnunet_plans": plans_bytes})
+        new_fit_cfg_fn = add_items_to_config_fn(self.strategy.configure_fit, {"nnunet_plans": self.nnunet_plans_bytes})
+        new_eval_cfg_fn = add_items_to_config_fn(
+            self.strategy.configure_evaluate, {"nnunet_plans": self.nnunet_plans_bytes}
+        )
         setattr(self.strategy, "configure_fit", new_fit_cfg_fn)
         setattr(self.strategy, "configure_evaluate", new_eval_cfg_fn)
 
         # Finish
         self.initialized = True
         log(INFO, "")
+
+    def save_server_state(self) -> None:
+        """
+        Save server checkpoint consisting of model, history, server round, metrics reporter and server name.
+            This method overrides parent to also checkpoint nnunet_plans, num_input_channels,
+            num_segmentation_heads and enable_deep_supervision.
+        """
+
+        assert self.per_round_checkpointer is not None
+
+        assert (
+            self.nnunet_plans_bytes is not None
+            and self.num_input_channels is not None
+            and self.num_segmentation_heads is not None
+            and self.enable_deep_supervision is not None
+            and self.nnunet_config is not None
+        )
+
+        ckpt = {
+            "model": self.server_model,
+            "history": self.history,
+            "current_round": self.current_round,
+            "metrics_reporter": self.metrics_reporter,
+            "server_name": self.server_name,
+            "nnunet_plans_bytes": self.nnunet_plans_bytes,
+            "num_input_channels": self.num_input_channels,
+            "num_segmentation_heads": self.num_segmentation_heads,
+            "enable_deep_supervision": self.enable_deep_supervision,
+            "nnunet_config": self.nnunet_config,
+        }
+
+        self.per_round_checkpointer.save_checkpoint(ckpt)
+
+        log(INFO, f"Saving server state to checkpoint at {self.per_round_checkpointer.checkpoint_path}")
+
+    def load_server_state(self) -> None:
+        """
+        Load server checkpoint consisting of model, history, server name, current round and metrics reporter.
+            The method overrides parent to add any necessary state when loading the checkpoint.
+        """
+        assert self.per_round_checkpointer is not None and self.per_round_checkpointer.checkpoint_exists()
+
+        ckpt = self.per_round_checkpointer.load_checkpoint()
```

Comment:
This is essentially equivalent to the `narrow_config_type` function in our utils/config.py file. Perhaps we generalize it to address any Dict?

Reply:
Yeah I guess the main difference is that the value is set as an attribute as opposed to returned. But I agree that it can be generalised. The solution would then be something like:
```python
setattr(self, name, narrow_dict_type(name, type, dict))
```

Reply:
Right, I missed that part. I think it makes sense to have a separate function that does narrowing and attribute setting

Reply:
alternatively the seperate function could inherit from the other one. Ie.

```python
def load_state_param(name, type, ckpt):
    val = narrow_dict_type(name, type, ckpt)
    setattr(self, name, val)
```

Not sure if adding a dependency would be a good or bad thing here

Reply:
Changed narrow_config_type to narrow_dict_type and added narrow_dict_type_and_set_attribute as per suggestions